server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -265,27 +271,60 @@ public CachingClusteredClient(
       }
 
       final Set<ServerToSegment> segments = computeSegmentsToQuery(timeline);
-      @Nullable final byte[] queryCacheKey = computeQueryCacheKey();
+      @Nullable
+      final byte[] queryCacheKey = computeQueryCacheKey();
+      @Nullable
+      final String queryResultKey = computeCurrentEtag(segments, queryCacheKey);
       if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) != null) {
-        @Nullable final String prevEtag = (String) query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH);
-        @Nullable final String currentEtag = computeCurrentEtag(segments, queryCacheKey);
+        @Nullable
+        final String prevEtag = (String) query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH);
+        @Nullable
+        final String currentEtag = queryResultKey;
         if (currentEtag != null && currentEtag.equals(prevEtag)) {
           return Sequences.empty();
         }
       }
 
+      @Nullable
+      final byte[] cachedResultSet = fetchFromResultLevelCache(queryResultKey);
+      if (cachedResultSet != null) {
+        log.info("Fetching entire result set from cache");
+        return fetchSequenceFromResultLevelCache(cachedResultSet);
+      }
       final List<Pair<Interval, byte[]>> alreadyCachedResults = pruneSegmentsWithCachedResults(queryCacheKey, segments);
       final SortedMap<DruidServer, List<SegmentDescriptor>> segmentsByServer = groupSegmentsByServer(segments);
-      return new LazySequence<>(() -> {
+      return Sequences.wrap(new LazySequence<>(() -> {
         List<Sequence<T>> sequencesByInterval = new ArrayList<>(alreadyCachedResults.size() + segmentsByServer.size());
         addSequencesFromCache(sequencesByInterval, alreadyCachedResults);
         addSequencesFromServer(sequencesByInterval, segmentsByServer);
         return Sequences
             .simple(sequencesByInterval)
             .flatMerge(seq -> seq, query.getResultOrdering());
+      }).map(r -> cacheResultEntry(r, queryResultKey)), new SequenceWrapper()
+      {
+        @Override
+        public void after(boolean isDone, Throwable thrown) throws Exception
+        {
+          final CachePopulator cachePopulator =
+              getResultCachePopulator(queryResultKey);
+          if (cachePopulator != null) {
+            cachePopulator.populate();
+          }
+        }
       });
     }
 
+    private T cacheResultEntry(T result, String queryResultKey)
```

Comment:
as per the other methods, this should return null immediately if queryResultKey is null

Reply:
The resultcachepopulator entry will be created only if the result cache is enabled and this method will add future entries only if the populator entry is present. This method would be harmless if resultcache is disabled.